firmware-support/bittide-sys/src/time.rs

```diff
@@ -0,0 +1,353 @@
+            panic!("Tried to create an Instant with a frequency of 0 hertz.")
+        }
+        Instant {
+            clock_cycles: (millis * frequency / 1e3 as u64),
+            frequency,
+        }
+    }
+
+    /// Create a new `Instant` from a number of seconds and the frequency of the timing peripheral.
+    pub fn from_secs<T: Into<u64>>(secs: T, frequency: u64) -> Instant {
+        if frequency == 0 {
+            panic!("Tried to create an Instant with a frequency of 0 hertz.")
+        }
+        Instant {
+            clock_cycles: secs.into() * frequency,
+            frequency,
+        }
+    }
+
+    /// The number of whole hours represented by this `Instant`.
+    pub fn to_hours(&self) -> u64 {
+        self.clock_cycles / (60 * 60 * self.frequency)
+    }
+
+    /// The number of whole minutes represented by this `Instant`.
+    pub fn to_mins(&self) -> u64 {
+        self.clock_cycles / (60 * self.frequency)
+    }
+
+    /// The number of whole seconds represented by this `Instant`.
+    pub fn to_secs(&self) -> u64 {
+        self.clock_cycles / self.frequency
+    }
+    /// The number of whole milliseconds represented by this `Instant`.
+    pub fn to_millis(&self) -> u64 {
+        (self.clock_cycles * 1e3 as u64) / self.frequency
+    }
+
+    /// The number of whole microseconds represented by this `Instant`.
+    pub fn to_micros(&self) -> u64 {
+        (self.clock_cycles * 1e6 as u64) / self.frequency
+    }
+
+    /// The number of cycles stored by this `Instant`.
+    pub fn get_cycles(&self) -> u64 {
+        self.clock_cycles
+    }
+
+    /// The frequency corresponding to this `Instant`s clock cycles.
+    pub fn get_frequency(&self) -> u64 {
+        self.frequency
+    }
+}
+
+impl ops::Add<Duration> for Instant {
+    type Output = Instant;
+
+    fn add(self, rhs: Duration) -> Instant {
+        Instant {
+            clock_cycles: self.clock_cycles + rhs.to_cycles(self.frequency),
+            frequency: self.frequency,
+        }
+    }
+}
+
+impl ops::AddAssign<Duration> for Instant {
+    fn add_assign(&mut self, rhs: Duration) {
+        self.clock_cycles += rhs.to_cycles(self.frequency)
+    }
+}
+
+impl ops::Sub<Duration> for Instant {
+    type Output = Instant;
+
+    fn sub(self, rhs: Duration) -> Instant {
+        Instant {
+            clock_cycles: self.clock_cycles - rhs.to_cycles(self.frequency),
+            frequency: self.frequency,
+        }
+    }
+}
+
+impl ops::SubAssign<Duration> for Instant {
+    fn sub_assign(&mut self, rhs: Duration) {
+        self.clock_cycles -= rhs.to_cycles(self.frequency)
+    }
+}
+
+impl ops::Sub<Instant> for Instant {
+    type Output = Duration;
+
+    fn sub(self, rhs: Instant) -> Duration {
+        Duration::from_micros(self.to_micros() - rhs.to_micros())
+    }
+}
+
+/// A representation of a relative time, stored in microseconds.
+#[derive(uDebug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Duration {
+    nanos: u64,
```

Comment:
```suggestion
    nanos: u64,
```

IMO either `ns` or `nanoseconds`, I don't think I've ever seen it abbreviated as `nanos`.